zilliqa/src/consensus.rs

```diff
@@ -1028,6 +1037,27 @@ impl Consensus {
         Ok(Some((proposal, broadcasted_transactions)))
     }
 
+    pub fn propose_pending_block(&mut self) -> Result<Block> {
+        let (proposal, applied_transactions) = self
+            .propose_block_inner()?
+            .ok_or_else(|| anyhow!("Unable to propose new block!"))?;
+
+        // Check if the votes entry is empty - if so - remove it
+        if let Some((signatures, ..)) = self.votes.get(&proposal.hash()) {
+            if signatures.is_empty() {
+                self.votes.remove(&proposal.hash());
+            }
+        }
+
+        // Re-insert all transactions that were executed during block proposal - they were removed from tx_pool and have to be re-added
+        for tx in applied_transactions {
```

Comment:
I think the way to do this is to have `propose_block_inner()` take a flag which makes it remember to not remove txns from the pool (or indeed do any other mutations), rather than have the code "just remember what propose_block_inner() does and undo it here" - that means there are two sources of truth - propose_block_inner() and this loop, and inevitably they will drift out of alignment with each other, leading to bugs like "if you call eth_getBalance() with block=pending, and you happen to hit the same node in which there are pending transactions, some of them will be dropped". I'm not sure I want to debug that!